multidownloader/evm_multidownloader_syncers.go

```diff
@@ -0,0 +1,103 @@
+	finality aggkittypes.BlockNumberFinality) (uint64, error) {
+	return dh.blockNotifierManager.GetCurrentBlockNumber(ctx, finality)
+}
+
+// BlockHeader gets the block header for the given 'block name'
+func (dh *EVMMultidownloader) BlockHeader(ctx context.Context,
+	finality aggkittypes.BlockNumberFinality) (*aggkittypes.BlockHeader, error) {
+	number, err := dh.blockNotifierManager.GetCurrentBlockNumber(ctx, finality)
+	if err != nil {
+		return nil, fmt.Errorf("EVMMultidownloader.BlockHeader: cannot get block number for finality=%s: %w",
+			finality.String(), err)
+	}
+	header, err := dh.ethClient.HeaderByNumber(ctx, big.NewInt(int64(number)))
+	if err != nil {
+		return nil, fmt.Errorf("EVMMultidownloader.BlockHeader: cannot get header for block number=%d: %w",
+			number, err)
+	}
+	return aggkittypes.NewBlockHeaderFromEthHeader(header), nil
+}
+
+// FilterLogs filters the logs. It gets it from storage or waits until they are available
+func (dh *EVMMultidownloader) FilterLogs(ctx context.Context, query ethereum.FilterQuery) ([]types.Log, error) {
+	dh.log.Debugf("EVMMultidownloader.FilterLogs: received query: %+v", query)
+	defer dh.log.Debugf("EVMMultidownloader.FilterLogs: finished query: %+v", query)
+	logQuery := mdrtypes.NewLogQueryFromEthereumFilter(query)
+	for !dh.IsAvailable(logQuery) {
+		dh.log.Infof("EVMMultidownloader.FilterLogs: waiting %s for logs to be available: %s",
+			dh.cfg.WaitPeriodToCheckCatchUp.String(), logQuery.String())
+		select {
+		case <-time.After(dh.cfg.WaitPeriodToCheckCatchUp.Duration):
+		case <-ctx.Done():
+			return nil, fmt.Errorf("EVMMultidownloader.FilterLogs: "+
+				"context done while waiting for logs %s to be available: %w",
+				logQuery.String(), ctx.Err())
+		}
+	}
+	logs, err := dh.storage.GetEthLogs(nil, logQuery)
+	if err != nil {
+		return nil, fmt.Errorf("EVMMultidownloader.FilterLogs: cannot get logs: %w", err)
+	}
+	dh.log.Debugf("EVMMultidownloader.FilterLogs(%d - %d): len(logs)= %d", query.FromBlock, query.ToBlock, len(logs))
+
+	return logs, nil
+}
+
+// HeaderByNumber gets the block header for the given block number from storage or ethClient
+func (dh *EVMMultidownloader) HeaderByNumber(ctx context.Context, number *big.Int) (*aggkittypes.BlockHeader, error) {
+	dh.log.Debugf("EVMMultidownloader.HeaderByNumber: received number: %s", number.String())
+	defer dh.log.Debugf("EVMMultidownloader.HeaderByNumber: finished number: %s", number.String())
+	if number.Cmp(big.NewInt(0)) < 0 {
+		return nil, fmt.Errorf("EVMMultidownloader.HeaderByNumber: negative block number not supported=%s", number.String())
+	}
+
+	block, _, err := dh.storage.GetBlockHeaderByNumber(nil, number.Uint64())
+	if err != nil {
+		return nil, fmt.Errorf("EVMMultidownloader.HeaderByNumber: cannot get BlockHeader number=%d: %w",
+			number.Uint64(), err)
+	}
+	if block != nil {
+		return block, nil
+	}
+	// This is a fallback mechanism in case the block is not found in storage (must be on storage!)
```

Comment:
If it must be on storage, should we just return an error in this case, or at least, store this fallback result to `db`?